### PR TITLE
Add kustomize overlays

### PR DESCRIPTION
**Pull request by tydanny**
## Pull request checklist

```release-notes

```

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature Enhancement
- [ ] Test Automation 
- [ ] Code Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Community Operator listing 
- [ ] Other (please describe): 

## What is the current behavior?

One default directory contains a default kustomization

## What is the new behavior?

The default overlay has been replaced with an overlays directory that allows multiple kustomizations.

## How risky is this change?
- [x] Small, isolated change
- [ ] Medium, requires regression testing
- [ ] Large, requires functional and regression testing

### CODE REVIEW

**Comment by nitishkumar4** [comment]
@tychristensen can you please update the file `config/manifests/kustomization.yaml` as well with `../overlays/default` to ensure `make bundle` works.

**Review by nitishkumar4 — CHANGES_REQUESTED** [review]
@tychristensen can you please update the file `config/manifests/kustomization.yaml` as well with `../overlays/default` to ensure `make bundle` works.

**Comment by tychristensen** [comment]
@nitishkumar4 I fixed that kustomization and altered the kustomize command in the bundle target so that it would override the image correctly. I think make bundle is working normally now but didn't any thorough testing.

**Comment by nitishkumar4** [comment]
Thanks for making changes @tychristensen. Should we change `IMAGE_TAG_BASE ?= quay.io/ibm-spectrum-scale/ibm-spectrum-scale-csi-operator` to `IMAGE_TAG_BASE ?= cp.icr.io/cp/spectrum/scale/csi/ibm-spectrum-scale-csi-operator` as well in `/operator/Makefile`

**Comment by tychristensen** [comment]
We should not change `IMAGE_TAG_BASE` since you guys want quay when you GA (or at least that was my understanding).

`go vet` is failing with the following message:

```bash
../../go/pkg/mod/sigs.k8s.io/json@v0.0.0-20211020170558-c049b76a60c6/internal/golang/encoding/json/encode.go:1249:12: sf.IsExported undefined (type reflect.StructField has no field or method IsExported)
../../go/pkg/mod/sigs.k8s.io/json@v0.0.0-20211020170558-c049b76a60c6/internal/golang/encoding/json/encode.go:1255:18: sf.IsExported undefined (type reflect.StructField has no field or method IsExported)
```

I think this might be because this requires golang 1.17 and the operator is using golang 1.16. I'm still investigating what needs to be done to resolve this.

**Comment by tychristensen** [comment]
I have fixed this PR so that it is now using the 5.1.3 RC2 images.

@madhuthorat I will be removing the overrides from CNSA soon (not sure when). If you need CNSA to be using specific images please use operator/config/overlays/cnsa/kustomization.yaml to specify the image locations/digests.

**Comment by tychristensen** [comment]
@nitishkumar4 I included the changes for `make manifests generate bundle` but it looks like there is something wrong with the go module that is preventing `make` from running successfully. I think this should be addressed in a separate PR because it would be to many unrelated changes to include in this PR.

File: operator/config/overlays/cnsa/kustomization.yaml [hunk 1]
```diff
@@ -0,0 +1,49 @@
+---
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+
+# Adds namespace to all resources.
+namespace: ibm-spectrum-scale-csi-driver
+
+bases:
+- ../../manager/
```
Comment (sam6258) [inline operator/config/overlays/cnsa/kustomization.yaml:9]: Shouldn't we just be building `default` and then overlay that?

Comment (sam6258) [reply operator/config/overlays/cnsa/kustomization.yaml:9]: Oh, looks like the way you have default kustomization file, we shouldn't do that.

**Review by nitishkumar4 — DISMISSED** [review]

File: operator/config/overlays/cnsa/kustomization.yaml (hunk 1 above)
Comment (tychristensen) [reply operator/config/overlays/cnsa/kustomization.yaml:9]: So we should leave the default at config/default then have a separate overlay directory that lists the default as a resource?

Comment (sam6258) [reply operator/config/overlays/cnsa/kustomization.yaml:9]: I thought you were gonna keep `config/default` in place when you mentioned overlays. But this is also fine I think.

File: generated/installer/ibm-spectrum-scale-csi-operator-dev.yaml [hunk 2]
```diff
@@ -33,18 +33,18 @@ spec:
       serviceAccountName: ibm-spectrum-scale-csi-operator
       containers:
       - name: operator
-        image: quay.io/ibm-spectrum-scale-dev/ibm-spectrum-scale-csi-operator:dev
+        image: quay.io/ibm-spectrum-scale-dev/ibm-spectrum-scale-csi-driver@sha256:241ab702e62c087b99eb5f2416675bc09b0f1d96365c34b3f292a6ca1429a9e7
```
Comment (nitishkumar4) [inline generated/installer/ibm-spectrum-scale-csi-operator-dev.yaml:36]: Should be operator image here.

File: generated/installer/ibm-spectrum-scale-csi-operator.yaml [hunk 3]
```diff
@@ -33,18 +33,18 @@ spec:
       serviceAccountName: ibm-spectrum-scale-csi-operator
       containers:
       - name: operator
-        image: cp.icr.io/cp/spectrum/scale/csi/ibm-spectrum-scale-csi-operator@sha256:241ab702e62c087b99eb5f2416675bc09b0f1d96365c34b3f292a6ca1429a9e7
+        image: quay.io/ibm-spectrum-scale-dev/ibm-spectrum-scale-csi-driver@sha256:241ab702e62c087b99eb5f2416675bc09b0f1d96365c34b3f292a6ca1429a9e7
```
Comment (nitishkumar4) [inline generated/installer/ibm-spectrum-scale-csi-operator.yaml:36]: should be operator image name.

**Comment by nitishkumar4** [comment]
> @nitishkumar4 I included the changes for `make manifests generate bundle` but it looks like there is something wrong with the go module that is preventing `make` from running successfully. I think this should be addressed in a separate PR because it would be to many unrelated changes to include in this PR.

Using Go1.17 should fix this.

File: operator/config/overlays/cnsa/kustomization.yaml [hunk 4]
```diff
@@ -0,0 +1,49 @@
+---
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+
+# Adds namespace to all resources.
+namespace: ibm-spectrum-scale-csi-driver
```
Comment (sam6258) [inline operator/config/overlays/cnsa/kustomization.yaml:6]: This namespace needs to be `ibm-spectrum-scale-csi`

Comment (sam6258) [reply operator/config/overlays/cnsa/kustomization.yaml:6]: I guess technically it doesn't have to be, but its one less thing CNSA has to worry about in its kustomization.yaml

Comment (sam6258) [reply operator/config/overlays/cnsa/kustomization.yaml:6]: Should we add `namespace.yaml` from CNSA repo to this overlay as well?

**Comment by deeghuge** [comment]
Closing as Amol Raised #715 for the same

**Pull request closed without merging**